nntrainer/src/databuffer.cpp

```diff
@@ -100,171 +101,115 @@ bool DataBuffer::init(int mini_batch, unsigned int train_bufsize, unsigned int v
 }
 
 void DataBuffer::UpdateData(buffer_type type, std::ifstream &file) {
+  unsigned int max_size = 0;
+  unsigned int buf_size = 0;
+  unsigned int *rest_size = NULL;
+  unsigned int *cur_size = NULL;
+  bool *running = NULL;
+  std::vector<std::vector<float>> *data = NULL;
+  std::vector<std::vector<float>> *datalabel = NULL;
+
   switch (type) {
-    case BUF_TRAIN: {
-      std::vector<unsigned int> mark;
-      mark.resize(max_train);
-      file.seekg(0, std::ios_base::end);
-      int64_t file_length = file.tellg();
+    case BUF_TRAIN:
+      max_size = max_train;
+      buf_size = train_bufsize;
+      rest_size = &rest_train;
+      cur_size = &cur_train_bufsize;
+      running = &train_running;
+      data = &trainData;
+      datalabel = &trainDataLabel;
+      break;
+    case BUF_VAL:
+      max_size = max_val;
+      buf_size = val_bufsize;
+      rest_size = &rest_val;
+      cur_size = &cur_val_bufsize;
+      running = &val_running;
+      data = &valData;
+      datalabel = &valDataLabel;
+      break;
+    case BUF_TEST:
+      max_size = max_test;
+      buf_size = test_bufsize;
+      rest_size = &rest_test;
+      cur_size = &cur_test_bufsize;
+      running = &test_running;
+      data = &testData;
+      datalabel = &testDataLabel;
+      break;
+    default:
+      break;
+  }
 
-      for (unsigned int i = 0; i < max_train; ++i) {
-        mark[i] = i;
-      }
+  unsigned int I;
+  std::vector<unsigned int> mark;
+  mark.resize(max_size);
+  file.seekg(0, std::ios_base::end);
+  int64_t file_length = file.tellg();
 
-      unsigned int I;
-      while (train_running && mark.size() != 0) {
-        if (train_bufsize - cur_train_bufsize > 0 && rest_train > 0) {
-          data_lock.lock();
-          std::vector<float> vec;
-          std::vector<float> veclabel;
+  for (unsigned int i = 0; i < max_size; ++i) {
+    mark[i] = i;
+  }
 
-          unsigned int id = rangeRandom(0, mark.size() - 1);
-          I = mark[id];
-          if (I > max_test)
-            throw std::runtime_error("Error: Test case id cannot exceed maximum number of test");
+  while ((*running) && mark.size() != 0) {
+    if (buf_size - (*cur_size) > 0 && (*rest_size) > 0) {
+      std::vector<float> vec;
+      std::vector<float> veclabel;
 
-          mark.erase(mark.begin() + id);
+      unsigned int id = rangeRandom(0, mark.size() - 1);
+      I = mark[id];
+      if (I > max_size)
+        ml_loge("Error: Test case id cannot exceed maximum number of test");
 
-          int64_t position = (I * input_size + I * class_num) * sizeof(float);
+      mark.erase(mark.begin() + id);
+      int64_t position = (I * input_size + I * class_num) * sizeof(float);
 
-          if (position > file_length)
-            throw std::runtime_error("Error: Cannot exceed max file size");
+      if (position > file_length)
```

Comment:
IMHO, if position is overflowed, it could not guard the bad things:(

Reply:
fixed. Thanks.